apis/logging/v1/output_types.go

```diff
@@ -276,6 +276,21 @@ type Splunk struct {
 	// Should be a valid JSON object
 	// +optional
 	Fields []string `json:"fields,omitempty"`
+
+	// IndexKey is a meta-data key field to use to send events to.
+	// For example: 'IndexKey: kubernetes.namespace_name` will use the kubernetes
+	// namespace as the index.
+	// If the IndexKey is not found, the `log_type` will be used as the index.
+	// Only one of IndexKey or IndexName can be defined.
+	// If IndexKey && IndexName are not specified, the default index defined within Splunk is used.
+	// +optional
+	IndexKey string `json:"indexKey,omitempty"`
+
+	// IndexName is the name of the index to send events to.
+	// Only one of IndexKey or IndexName can be defined.
+	// If IndexKey && IndexName are not specified, the default index defined within Splunk is used.
+	// +optional
+	IndexName string `json:"indexName,omitempty"`
```

Comment:
@alanconway this is consistent with our Elasticsearch output but I'm not certain it is needed.  We could embrace the notion of using curly brackets to differentiate dynamic versus static like `{{kubernetes.namespace_name}}` and have a single field.  

Do we leave as-is for now and revisit in the next version of the API?

Reply:
I like the idea of using {{}} instead of separate Key+Name fields, but we should make the same change to other places we have Key/Name pairs like this. We could make the change backwards compat now by having
- Index: index string, may include field templates such as {{kuberntes.namespace_name}}
- IndexName: literal index string
- IndexKey: name of a field to use as the index string. Note: `indexKey: "foo"` is equivalent to `index:{{foo}}`
and deprecate IndexName/IndexKey for removal in v2 APIs.

Reply:
> we should make the same change to other places

I think Elasticsearch 'structuredKey' and 'structuredName' is the only other pairing.  We could introduce 'index' there with the same convention you identified, excluding internally managed ES

Reply:
If using `Index`, if it includes a field that is not available what should the default behavior be?

For example,
I define this as an index: `{{kubernetes.field}}_customIndex`
Log 1 has `kubernetes.field = foo` but log 2 doesn't include `kubernetes.field`

Log 1's index will now be `foo_customIndex` but what should log 2 be?
1. Invalid? Write to `default` splunk index?
2. Omit only `kubernetes.field` and write to `_customIndex`?

Reply:
> Log 1's index will now be foo_customIndex but what should log 2 be?
>
> Invalid? Write to default splunk index?
>2.  Omit only kubernetes.field and write to _customIndex?


@alanconway @cahartma I lean towards documenting absence implies `""` and the result is as suggested by 2.  It would be path of least surprise.  Additionally, IMO, defining a custom index in this way implies a schema; you know something about your log data and want it organized consistently.  This question is fundamentally the issue we have always had with log data not having any consistency.  

I can imagine this issue will pop up when:
* you send journal/audit logs through the same pipeline as container

Reply:
As discussed, using an index with templating will be lined up for the next release